mqtt/mqtt-broker/src/edgehub_auth.rs

```diff
@@ -0,0 +1,370 @@
+    identity: Option<String>,
+}
+
+impl EdgeHubAuthResponse {
+    fn into_auth_id(self) -> Result<Option<AuthId>, Error> {
+        match self.result {
+            EdgeHubResultCode::Authenticated => self
+                .identity
+                .map_or(Err(ErrorKind::AuthProcessResponse.into()), |i| {
+                    Ok(Some(AuthId::Value(i)))
+                }),
+            EdgeHubResultCode::Unauthenticated => Ok(None),
+        }
+    }
+}
+
+macro_rules! impl_for{
+    ($fun_name:ident, $typ:ident) => {
+        fn $fun_name<E: serde::de::Error>(self, v: $typ) -> Result<EdgeHubResultCode, E> {
+            match v {
+                200 => Ok(EdgeHubResultCode::Authenticated),
+                403 => Ok(EdgeHubResultCode::Unauthenticated),
+                v => Err(E::custom(format!("unexpected int: {}", v))),
+            }
+        }
+    }
+}
+
+struct EdgeHubResultCodeVisitor;
+impl<'de> serde::de::Visitor<'de> for EdgeHubResultCodeVisitor {
+    type Value = EdgeHubResultCode;
+    fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("200 or 403")
+    }
+
+    impl_for!(visit_i32, i32);
+    impl_for!(visit_i64, i64);
+    impl_for!(visit_u32, u32);
+    impl_for!(visit_u64, u64);
+}
+
+impl<'de> serde::Deserialize<'de> for EdgeHubResultCode {
+    fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<EdgeHubResultCode, D::Error> {
+        d.deserialize_i32(EdgeHubResultCodeVisitor)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use base64::decode;
+    use matches::assert_matches;
+    use mockito::{mock, Matcher};
+
+    use crate::auth::{AuthId, Authenticator, Certificate, Credentials};
+    use crate::edgehub_auth::EdgeHubAuthenticator;
+    use crate::Error;
+
+    const CERT: &str = "MIIBLjCB1AIJAOTg4Zxl8B7jMAoGCCqGSM49BAMCMB8xHTAbBgNVBAMMFFRodW1i\
+                        cHJpbnQgVGVzdCBDZXJ0MB4XDTIwMDQyMzE3NTgwN1oXDTMzMTIzMTE3NTgwN1ow\
+                        HzEdMBsGA1UEAwwUVGh1bWJwcmludCBUZXN0IENlcnQwWTATBgcqhkjOPQIBBggq\
+                        hkjOPQMBBwNCAARDJJBtVlgM0mBWMhAYagF7Wuc2aQYefhj0cG4wAmn3M4XcxJ39\
+                        XkEup2RRAj7SSdOYhTmRpg5chhpZX/4/eF8gMAoGCCqGSM49BAMCA0kAMEYCIQD/\
+                        wNzMjU1B8De5/+jEif8rkLDtqnohmVRXuAE5dCfbvAIhAJTJ+Fyg19uLSKVyOK8R\
+                        5q87sIqhJXhTfNYvIt77Dq4J";
+
+    #[tokio::test]
+    async fn test_call_auth_agent() {
+        let _m = mock("POST", "/authenticate/")
+        .with_status(200)
+        .with_body(r#"{"result": 200, "identity":"vikauthtest.azure-devices.net/ca_root","version": "2020-04-20"}"#)
+        .create();
+
+        let authenticator = EdgeHubAuthenticator {
+            url: mockito::server_url() + "/authenticate/",
+        };
+        let cert_as_string = CERT.to_string();
+        let raw_cert_result = decode(cert_as_string);
+        let raw_cert = raw_cert_result.unwrap();
+        let cert: Certificate = Certificate::from(raw_cert);
```

Comment:
if you use from you dont need to put type for definition
```suggestion
        let cert = Certificate::from(raw_cert);
```